### PR TITLE
0.5: Revise exposed crate features

### PR DESCRIPTION
**Pull request by pitdicker**
Remove the old crate features `libc`, `old_time`, `rkyv` and `winapi`.
Switch to the `dep:` syntax for optional dependencies that where exposed as features: `pure-rust-locales`, `android-tzdata`, `iana-time-zone`, `js-sys`, `wasm-bindgen` and `windows-targets`.

### CODE REVIEW

**Comment by pitdicker** [comment]
Also updated the description in `lib.rs` and the readme, and removed the `__unstable_bench` feature.

The bench feature was only used for benchmarking `YearFlags::from_year`. All that method does is:
```rust
let year = year.rem_euclid(400);
YEAR_TO_FLAGS[year as usize]
```

A division and a table lookup doesn't seem all that interesting to benchmark, let alone have a feature for.

File: Cargo.toml [hunk 1]
```diff
@@ -16,31 +16,38 @@ rust-version = "1.61.0"
...
 
-[features]
-# Don't forget to adjust `ALL_NON_EXCLUSIVE_FEATURES` in CI scripts when adding a feature or an optional dependency.
-default = ["clock", "std"]
-alloc = []
-libc = []
-winapi = ["windows-targets"]
-std = ["alloc"]
-clock = ["winapi", "iana-time-zone", "android-tzdata", "now"]
+[features] # Don't forget to adjust `ALL_NON_EXCLUSIVE_FEATURES` in CI scripts when adding a feature or an optional dependency.
```
Comment (djc) [inline Cargo.toml:19]: This commit has a lot going on, can you split it in smaller pieces? For example, one for introducing the use of `dep:`, one for cleaning up the `clock`-related stuff, other stuff...

Also I think I'd prefer to have single-line comments on the same line as the feature, and keep lists of features on one line.

Comment (pitdicker) [reply Cargo.toml:19]: Done. Second try on making the commits somewhat reasonable :smile:.

> Also I think I'd prefer to have single-line comments on the same line as the feature, and keep lists of features on one line.

Sorry, I didn't write down what I was trying to do. libs.rs can show documentation for crate features but I am not sure of the details. Having a comment above the feature works for sure.

We have a CI run that checks the format of cargo.toml. I think that is quite useless personally. It wants the list of features for `clock` to be on multiple lines.

File: Cargo.toml [hunk 2]
```diff
@@ -18,16 +18,16 @@ name = "chrono"
 
 [features]
 # Don't forget to adjust `ALL_NON_EXCLUSIVE_FEATURES` in CI scripts when adding a feature or an optional dependency.
-default = ["clock", "std"]
-alloc = []
-std = ["alloc"]
+default = ["clock"]
```
Comment (djc) [inline Cargo.toml:20]: Not sure about this change? In our code, do we actually only need `std` for things that are related to `clock`? Or is this more of an incidental dependency rather than a conceptual one?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -87,7 +87,6 @@ jobs:
       - run: cargo test --no-default-features --features=alloc
       - run: cargo test --no-default-features --features=unstable-locales
       - run: cargo test --no-default-features --features=alloc,unstable-locales
-      - run: cargo test --no-default-features --features=now
```
Comment (djc) [inline .github/workflows/test.yml:92]: (Depends on the outcome of my other comment.)

**Comment by djc** [comment]
Much better, I think that split was definitely worth it!

File: Cargo.toml (hunk 2 above)
Comment (pitdicker) [reply Cargo.toml:20]: We used to have custom code for getting the current time from the OS but today we are only a wrapper around `SystemTime::now()`.

Comment (djc) [reply Cargo.toml:20]: I guess what I mean is, is that the only use of `std` in the library? I guess all the other stuff is just `alloc`?

Comment (pitdicker) [reply Cargo.toml:20]: We only use `std` for the `Error` trait, `SystemTime::now()`, and the stuff in the unix module.

Comment (djc) [reply Cargo.toml:20]: Okay, but our usage for the `Error` trait doesn't seem to be scoped to `clock` only, right?

Comment (pitdicker) [reply Cargo.toml]: I've removed the commit.

File: README.md [hunk 4]
```diff
@@ -50,12 +50,10 @@ See [docs.rs](https://docs.rs/chrono/latest/chrono/) for the API reference.
 
 Default features:
 
+* `clock`: Enables reading the local timezone (`Local`), superset of `now`.
+* `now`: Enables reading the system time (`now()`) using the standard library.
+* `std`: Implements `std::error::Error` for error types, superset of `alloc`.
```
Comment (djc) [inline README.md:55]: I'd like to avoid the "superset of" wording, instead using something like "implies" or "also enables". Should we also make explicit that `now` enables `std` here?

**Review by djc — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (djc) [reply README.md:55]: I think this is in firmly in [Hyrum's Law](https://www.hyrumslaw.com/) territory, and generally will be seen as semver-breaking anyway if we decide to go back on it so might as well document it.

Comment (pitdicker) [reply README.md]: Haha, yes that is something to worry about. But I don't think that will be a problem here.

I added a commit "Don't let `now` imply `std` feature". Do you want to have a look?

With it the `std` feature only enables the `Error` trait, `TryFrom<SystemTime>`, and methods that depend on `alloc`.

I had to extract the code from `TryFrom<SystemTime>` into a method on `DateTime<Utc>` to split the `now` feature from `std`.

Comment (djc) [reply README.md]: So what functionality does `now` enable when `std` is disabled?

Comment (pitdicker) [reply README.md]: Only `Utc::now()`. And the `clock` feature adds `Local::now()`.

Comment (djc) [reply README.md]: `Utc::now()` calls `SystemTime::now()`, so it's actually dependent on `std`, right?

Comment (pitdicker) [reply README.md]: Currently it does on all platforms.

Comment (pitdicker) [reply README.md]: Now I wonder if we shouldn't just remove the `now` feature and fold it into `std`.

Comment (djc) [reply README.md]: But on WASM we need other stuff for `now`, right?

Comment (pitdicker) [reply README.md]: You are completely right, I forgot. It wouldn't have to depend on std on that platform.

**Comment by pitdicker** [comment]
Sorry that this is taking so many back-and-fort between us. But I'm really starting to like the changes here as good cleanups.

File: Cargo.toml [hunk 5]
```diff
@@ -27,7 +27,7 @@ clock = [
   "dep:android-tzdata",
   "now",
 ]
-now = ["std"]
+now = []
```
Comment (djc) [inline Cargo.toml:33]: So I think we should revert this change, because `now` without `std` is pretty much meaningless.

Comment (djc) [reply Cargo.toml:33]: I think `now` should enable calling `Utc::now()` on most common platforms. If it doesn't, that feels broken.

Comment (pitdicker) [reply Cargo.toml:33]: It does. https://github.com/chronotope/chrono/pull/1552/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R474

Comment (djc) [reply Cargo.toml:33]: It seems surprising that the `cfg(feature = "std")` guard doesn't match the behavior of whether we enable/disable `no_std` exactly (for the library).

Comment (pitdicker) [reply Cargo.toml:33]: It's a choice we can make. Separating the features allows things like using `Utc::now` in combination with `wasmbind` without depending on the standard library.

Comment (djc) [reply Cargo.toml:33]: Okay, let's try it. Easier to add it later than to remove it.
